runners/direct-java/src/main/java/org/apache/beam/runners/direct/GroupAlsoByWindowEvaluatorFactory.java

```diff
@@ -272,44 +277,13 @@ public void outputWindowedValue(
     }
 
     @Override
-    public TimerInternals timerInternals() {
-      throw new UnsupportedOperationException(
-          String.format(
-              "%s should use the %s it is provided rather than the contents of %s",
-              ReduceFnRunner.class.getSimpleName(),
-              TimerInternals.class.getSimpleName(),
-              getClass().getSimpleName()));
-    }
-
-    @Override
-    public Collection<? extends BoundedWindow> windows() {
-      throw new IllegalArgumentException(
-          String.format(
-              "%s should not access Windows via %s.windows(); "
-                  + "it should instead inspect the window of the input elements",
-              GroupAlsoByWindowEvaluator.class.getSimpleName(),
-              WindowingInternals.class.getSimpleName()));
-    }
-
-    @Override
-    public PaneInfo pane() {
-      throw new IllegalArgumentException(
-          String.format(
-              "%s should not access Windows via %s.windows(); "
-                  + "it should instead inspect the window of the input elements",
-              GroupAlsoByWindowEvaluator.class.getSimpleName(),
-              WindowingInternals.class.getSimpleName()));
-    }
-
-    @Override
-    public <T> void writePCollectionViewData(
-        TupleTag<?> tag, Iterable<WindowedValue<T>> data, Coder<T> elemCoder) throws IOException {
-      throw new UnsupportedOperationException();
-    }
-
-    @Override
-    public <T> T sideInput(PCollectionView<T> view, BoundedWindow mainInputWindow) {
-      throw new UnsupportedOperationException();
+    public <SideOutputT> void sideOutputWindowedValue(
+        TupleTag<SideOutputT> tag,
+        SideOutputT output,
+        Instant timestamp,
+        Collection<? extends BoundedWindow> windows,
+        PaneInfo pane) {
+      throw new UnsupportedOperationException("Can't output to side outputs from a ReduceFn");
```

Comment:
Let's call it `GroupAlsoByWindow`, since actually that is the criteria. If `ReduceFn` ever becomes a thing, it might be allowed to side output. In fact, stateful `DoFn` is basically `ReduceFn`.